genePred/genePred.go

```diff
@@ -0,0 +1,166 @@
+package genePred
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/fileio"
+	"io"
+	"log"
+	"strings"
+)
+
+type GenePred struct {
+	Id         string
+	Symbol     string
+	Chrom      string
+	Strand     byte
+	TxStart    int
+	TxEnd      int
+	CdsStart   int
+	CdsEnd     int
+	ExonStarts []int
+	ExonEnds   []int
+	ExonFrames []int
+	Score      int
+}
+
+func GenePredToString(g *GenePred) string {
+	var answer string
+
+	answer = fmt.Sprintf("%s\t%s\t%s\t%s\t%v\t%v\t%v\t%v\t%s\t%s\t%s\t%v", g.Id, g.Symbol, g.Chrom, string(g.Strand), g.TxStart, g.TxEnd, g.CdsStart, g.CdsEnd, SliceIntToString(g.ExonStarts), SliceIntToString(g.ExonEnds), SliceIntToString(CalcExonFrame(g)), g.Score)
+	return answer
+}
+
+//WriteToFileHandle writes an input GenePred struct with a specified number of fields to an io.Writer
+func WriteToFileHandle(file io.Writer, records []*GenePred) {
+	for _, rec := range records { //take out if we need writeSliceToFileHandle
+		var err error
+		_, err = fmt.Fprintf(file, "%s\n", GenePredToString(rec))
+		//TODO: fmt.Fprintf is slow
+		common.ExitIfError(err)
+	}
+}
+
+//Write writes a slice of GenePred structs with a specified number of fields to a specified filename.
+func Write(filename string, records []*GenePred) {
+	file := fileio.EasyCreate(filename)
+	defer file.Close()
+
+	WriteToFileHandle(file, records)
+}
+
+func Read(filename string) []*GenePred {
+	var line string
+	var answer []*GenePred
+	var doneReading = false
+
+	file := fileio.EasyOpen(filename)
+	defer file.Close()
+
+	for line, doneReading = fileio.EasyNextRealLine(file); !doneReading; line, doneReading = fileio.EasyNextRealLine(file) {
+		current := processGenePredLine(line)
+		answer = append(answer, current)
+	}
+	return answer
+}
+
+func processGenePredLine(line string) *GenePred {
+	current := GenePred{}
+
+	words := strings.Split(line, "\t")
+	current.Id = words[0]
+	current.Symbol = words[0]
+	current.Chrom = words[1]
+	if words[2] == "+" {
+		current.Strand = '+'
+	} else if words[2] == "." {
+		current.Strand = '.'
+	} else if words[2] == "-" {
+		current.Strand = '-'
+	} else {
+		log.Fatal("no strand specified")
+	}
+	current.TxStart = common.StringToInt(words[3])
+	current.TxEnd = common.StringToInt(words[4])
+	current.CdsStart = common.StringToInt(words[5])
+	current.CdsEnd = common.StringToInt(words[6])
+	current.ExonStarts = StringToIntSlice(words[7])
```

Comment:
I think that the genePred file format will have the exon number as a column in the file after CdsEnd and before ExonStart.  This means that exonNumber from a few lines down would be set equal to "common.StringToInt(words[7])" and then words[8] would be processed to get ExonStarts.
To see an example genePred file, you can go to the ucsc genome browser, click on tools, click on table browser.  Now select a gene track and get the output as "all fields from selected table".  When you get the data it will be a genePred file.